protocol/binary_test.go

```diff
@@ -560,6 +561,66 @@ func TestBinaryEOFFailure(t *testing.T) {
 	}
 }
 
+type limitWriter struct {
+	b []byte
+}
+
+var errWriteLimitReached = errors.New("write limit reached")
+
+func (lw *limitWriter) Write(p []byte) (int, error) {
+	fmt.Printf("[limitWriter] write %v bytes\n", len(p))
+	n := copy(lw.b, p)
+	lw.b = lw.b[n:]
+	if n < len(p) {
+		return n, errWriteLimitReached
+	}
+	return n, nil
```

Comment:
Since it doesn't seem we're doing anything with the written bytes in the test case, could we just write the bytes to an [io.Discard](https://pkg.go.dev/io#Discard) and set a flag on on writer to direct it to error in the cases that you want it to? Is a limit writer a specific error case you're trying to exercise or will any error on the underlying Write do?

Reply:
WriteString internally calls the Write method twice sequentially:
1. for int32
2. for []byte
The second write will only be called if the first succeeds. We want to simulate a failure for both. 
Since there is a single `io.Writer`, I couldn't use a simpler failAlwaysWriter. 
A limit writer seemed fitting here as its errors are easy to reason about. It fails if we attempt to write more than what has been allotted. It could also be reused in other tests that require the Write call to fail. The advantage of doing an actual copy() instead of an `io.Discard` is that this implementation can also be used for testing non-error scenarios. I assumed its utility outweighs test performance.
Wdyt?

Reply:
Wanting to test both write cases makes sense, thanks for clarifying!

As for the copy() versus io.Discard, I do understand the potential usability. However, it took me some time to speculate on why live behavior like copy() was used when I didn't see the written bytes used anywhere in the PR -- I actively re-read the code multiple times in search for where the copy would be crucial. I think if I saw something like an io.Discard I would have interpreted the writer as simply testing error cases and would have not spent time on an additional search of the code.